main/pipeline/pipeline.cc

```diff
@@ -428,55 +428,107 @@ struct IndexThreadResultPack {
     ProgressIndicator progress(opts.showProgress, "Indexing", input->bound);
-    Timer timeit(cgs->tracer(), "mergeIndexResults");
-    IndexThreadResultPack threadResult;
-    IndexResult ret;
-    for (auto result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs->tracer()); !result.done();
-         result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs->tracer())) {
-        if (result.gotItem()) {
-            counterConsume(move(threadResult.counters));
-            if (ret.gs == nullptr) {
-                ret.gs = move(threadResult.res.gs);
-                ENFORCE(ret.trees.empty());
-                ret.trees = move(threadResult.res.trees);
-            } else {
-                core::GlobalSubstitution substitution(*threadResult.res.gs, *ret.gs, cgs.get());
-                {
-                    Timer timeit(cgs->tracer(), "substituteTrees");
-                    for (auto &tree : threadResult.res.trees) {
-                        auto file = tree.file;
-                        if (!file.data(*ret.gs).cached) {
-                            core::MutableContext ctx(*ret.gs, core::Symbols::root(), file);
-                            tree.tree = ast::Substitute::run(ctx, substitution, move(tree.tree));
+    Timer timeit(cgs.tracer(), "mergeIndexResults");
+
+    auto batchq = make_shared<ConcurrentBoundedQueue<IndexSubstitutionJob>>(input->bound);
+    vector<ast::ParsedFile> ret;
+
+    {
+        Timer timeit(cgs.tracer(), "mergeGlobalStates");
+        IndexThreadResultPack threadResult;
+        for (auto result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs.tracer());
+             !result.done(); result = input->wait_pop_timed(threadResult, WorkerPool::BLOCK_INTERVAL(), cgs.tracer())) {
+            if (result.gotItem()) {
+                counterConsume(move(threadResult.counters));
+                auto numTrees = threadResult.res.trees.size();
+                batchq->push(IndexSubstitutionJob{cgs, std::move(threadResult.res)}, numTrees);
+            }
+        }
+    }
+
+    {
+        Timer timeit(cgs.tracer(), "substituteTrees");
+        auto resultq = make_shared<BlockingBoundedQueue<vector<ast::ParsedFile>>>(batchq->bound);
+
+        workers.multiplexJob("substituteTrees", [&cgs, batchq, resultq]() {
+            Timer timeit(cgs.tracer(), "substituteTreesWorker");
+            IndexSubstitutionJob job;
+            for (auto result = batchq->try_pop(job); !result.done(); result = batchq->try_pop(job)) {
+                if (result.gotItem()) {
+                    if (job.subst.has_value()) {
+                        for (auto &tree : job.trees) {
+                            auto file = tree.file;
+                            if (!file.data(cgs).cached) {
+                                core::MutableContext ctx(cgs, core::Symbols::root(), file);
```

Comment:
Why does `ast::Substitute` need a `MutableContext`? Is this a mistake?

Reply:
I'm not sure, I just moved over what was there already. I'd be happy to experiment with a `core::Context` instead.

Reply:
`LazyGlobalSubstitution` requires a `core::MutableContext`, which is why `Substitute` is written assuming a `core::MutableSubstitution`.